include/picolibrary/wiznet/w5500/ip/tcp.h

```diff
@@ -23,10 +23,1320 @@
+            } // if
+
+            if ( end - begin > free_size ) {
+                end = begin + free_size;
+            } // if
+        }
+
+        SN_TX_WR::Type sn_tx_wr;
+        {
+            auto result = m_driver->read_sn_tx_wr( m_socket_id );
+            if ( result.is_error() ) {
+                return result.error();
+            } // if
+
+            sn_tx_wr = result.value();
+        }
+
+        {
+            auto result = m_driver->write( m_socket_id, sn_tx_wr, begin, end );
+            if ( result.is_error() ) {
+                return result.error();
+            } // if
+        }
+
+        {
+            auto result = m_driver->write_sn_tx_wr( m_socket_id, sn_tx_wr + ( end - begin ) );
+            if ( result.is_error() ) {
+                return result.error();
+            } // if
+        }
+
+        {
+            auto result = m_driver->write_sn_cr(
+                m_socket_id, static_cast<SN_CR::Type>( Command::SEND ) );
+            if ( result.is_error() ) {
+                return result.error();
+            } // if
+        }
+
+        for ( ;; ) {
+            auto result = m_driver->read_sn_cr( m_socket_id );
+            if ( result.is_error() ) {
+                return result.error();
+            } // if
+
+            if ( not result.value() ) {
+                m_is_transmitting = true;
+
+                return end;
+            } // if
+        }     // for
+    }
+
+    /**
+     * \brief Check if data transmission is in progress.
+     *
+     * \return true if data transmission is in progress.
+     * \return false if data transmission is not in progress.
+     */
+    constexpr auto is_transmitting() const noexcept
+    {
+        return m_is_transmitting;
+    }
+
+    /**
+     * \brief Manually transmit a keepalive packet.
+     *
+     * \return Nothing if transmitting the keepalive packet succeeded.
+     * \return picolibrary::Generic_Error::NOT_CONNECTED if the socket is not connected to
+     *         a remote endpoint.
+     * \return picolibrary::WIZnet::W5500::IP::Network_Stack::nonresponsive_device() if
+     *         the W5500 is nonresponsive.
+     * \return An error code if transmitting the keepalive packet failed for any other
+     *         reason.
+     */
+    auto transmit_keepalive() noexcept -> Result<Void, Error_Code>
+    {
+        // #lizard forgives the length
+
+        if ( m_state != State::CONNECTED ) {
+            return Generic_Error::NOT_CONNECTED;
+        } // if
+
+        {
+            auto result = m_driver->read_sn_sr( m_socket_id );
+            if ( result.is_error() ) {
+                return result.error();
+            } // if
+
+            switch ( static_cast<Socket_Status>( result.value() ) ) {
+                case Socket_Status::CLOSED: return Generic_Error::NOT_CONNECTED;
+                case Socket_Status::ESTABLISHED: break;
+                case Socket_Status::FIN_WAIT: return Generic_Error::WOULD_BLOCK;
```

Comment:
This function should not be called once a graceful shutdown is in progress. Return `Generic_Error::NOT_CONNECTED` instead of `Generic_Error::WOULD_BLOCK` for this socket status for better consistency.